activerecord/test/cases/insert_all_test.rb

```diff
@@ -336,6 +337,22 @@ def test_upsert_logs_message_including_model_name
     end
   end
 
+  def test_upsert_and_db_warnings
+    skip unless supports_insert_on_duplicate_update?
+
+    begin
+      with_db_warnings_action(:raise) do
```

Comment:
Not worth doing here, more for a followup, but I wonder if it would make sense to run CI/tests with DB warnings treated as errors by default. I think it might have caught this issue way sooner.

Also we probably should backport that fix.